ports/nrf/common-hal/bleio/Scanner.c

```diff
@@ -30,79 +31,76 @@
         return;
     }
 
-    // TODO: Don't add new entry for each item, group by address and update
     bleio_scanentry_obj_t *entry = m_new_obj(bleio_scanentry_obj_t);
     entry->base.type = &bleio_scanentry_type;
     entry->rssi = report->rssi;
 
-    entry->address.type = report->peer_addr.addr_type;
-    memcpy(entry->address.value, report->peer_addr.addr, BLEIO_ADDRESS_BYTES);
+    bleio_address_obj_t *address = m_new_obj(bleio_address_obj_t);
+    address->base.type = &bleio_address_type;
+    common_hal_bleio_address_construct(MP_OBJ_TO_PTR(address),
+                                       report->peer_addr.addr, report->peer_addr.addr_type);
+    entry->address = address;
 
-#if (BLUETOOTH_SD == 140)
-    entry->data = mp_obj_new_bytearray(report->data.len, report->data.p_data);
-#else
-    entry->data = mp_obj_new_bytearray(report->dlen, report->data);
-#endif
+    entry->data = mp_obj_new_bytes(report->data.p_data, report->data.len);
 
-    mp_obj_list_append(scanner->adv_reports, entry);
+    mp_obj_list_append(scanner->scan_entries, MP_OBJ_FROM_PTR(entry));
 
-#if (BLUETOOTH_SD == 140)
     const uint32_t err_code = sd_ble_gap_scan_start(NULL, &m_scan_buffer);
     if (err_code != NRF_SUCCESS) {
         mp_raise_OSError_msg_varg(translate("Failed to continue scanning, err 0x%04x"), err_code);
     }
-#endif
 }
 
-void common_hal_bleio_scanner_scan(bleio_scanner_obj_t *self, mp_int_t timeout) {
-    ble_drv_add_event_handler(on_ble_evt, self);
+void common_hal_bleio_scanner_construct(bleio_scanner_obj_t *self) {
+    self->scan_entries = mp_obj_new_list(0, NULL);
+}
+
+void common_hal_bleio_scanner_scan(bleio_scanner_obj_t *self, mp_float_t timeout, mp_float_t interval, mp_float_t window) {
+    common_hal_bleio_adapter_set_enabled(true);
+    ble_drv_add_event_handler(scanner_on_ble_evt, self);
 
     ble_gap_scan_params_t scan_params = {
-        .interval = MSEC_TO_UNITS(self->interval, UNIT_0_625_MS),
-        .window = MSEC_TO_UNITS(self->window, UNIT_0_625_MS),
-#if (BLUETOOTH_SD == 140)
+        .interval = SEC_TO_UNITS(interval, UNIT_0_625_MS),
+        .window = SEC_TO_UNITS(window, UNIT_0_625_MS),
         .scan_phys = BLE_GAP_PHY_1MBPS,
-#endif
     };
 
-    common_hal_bleio_adapter_set_enabled(true);
+    // Empty the advertising reports list.
+    mp_obj_list_clear(self->scan_entries);
 
     uint32_t err_code;
-#if (BLUETOOTH_SD == 140)
     err_code = sd_ble_gap_scan_start(&scan_params, &m_scan_buffer);
-#else
-    err_code = sd_ble_gap_scan_start(&scan_params);
-#endif
 
     if (err_code != NRF_SUCCESS) {
         mp_raise_OSError_msg_varg(translate("Failed to start scanning, err 0x%04x"), err_code);
     }
 
-    if (timeout > 0) {
-        mp_hal_delay_ms(timeout);
-        sd_ble_gap_scan_stop();
-    }
+    mp_hal_delay_ms(timeout * 1000);
+    sd_ble_gap_scan_stop();
+}
+
+mp_obj_t common_hal_bleio_scanner_get_scan_entries(bleio_scanner_obj_t *self) {
+    return self->scan_entries;
```

Comment:
Why not just return the scan entries from scan?

Reply:
`scan()` does return the entries. Do you mean why allow them to be fetched through an attribute too? Interesting question: there's no reason to make them fetchable later. I'll remove the attribute.

Reply:
I mean I'll just have the common_hal scan routine return them. I think this was left over from the old code.